File: code/client/munkilib/updatecheck/compare.py
# encoding: utf-8
#
# Copyright 2009-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
updatecheck.compare

Created by Greg Neagle on 2016-12-13.

Comparison/checking functions used by updatecheck
"""
from __future__ import absolute_import, print_function

import os
from operator import itemgetter

from .. import display
from .. import munkihash
from .. import info
from .. import pkgutils
from .. import utils
from .. import FoundationPlist


ITEM_DOES_NOT_MATCH = VERSION_IS_LOWER = -1
ITEM_NOT_PRESENT = 0
ITEM_MATCHES = VERSION_IS_THE_SAME = 1
VERSION_IS_HIGHER = 2


def compare_versions(thisvers, thatvers):
    """Compares two version numbers to one another.

    Returns:
      -1 if thisvers is older than thatvers
      1 if thisvers is the same as thatvers
      2 if thisvers is newer than thatvers
    """
    if (pkgutils.MunkiLooseVersion(thisvers) <
            pkgutils.MunkiLooseVersion(thatvers)):
        return VERSION_IS_LOWER
    elif (pkgutils.MunkiLooseVersion(thisvers) ==
          pkgutils.MunkiLooseVersion(thatvers)):
        return VERSION_IS_THE_SAME
    return VERSION_IS_HIGHER


def compare_application_version(app):
    """Checks the given path if it's available,
    otherwise uses LaunchServices and/or Spotlight to look for the app

    Args:
      app: dict with application bundle info

    Returns:
         0 if the app isn't installed
            or doesn't have valid Info.plist
        -1 if it's older
         1 if the version is the same
         2 if the version is newer

    Raises utils.Error if there's an error in the input
    """
    if 'path' in app:
        filepath = os.path.join(app['path'], 'Contents', 'Info.plist')
        if os.path.exists(filepath):
            return compare_bundle_version(app)
        display.display_debug2('%s doesn\'t exist.', filepath)
        return ITEM_NOT_PRESENT

    # no 'path' in dict
    display.display_debug2('No path given for application item.')
    # let's search:
    name = app.get('CFBundleName', '')
    bundleid = app.get('CFBundleIdentifier', '')
    version_comparison_key = app.get(
        'version_comparison_key', 'CFBundleShortVersionString')
    versionstring = app.get(version_comparison_key)

    if name == '' and bundleid == '':
        # no path, no name, no bundleid. Error!
        raise utils.Error(
            'No path, application name or bundleid was specified!')

    display.display_debug1(
        'Looking for application %s with bundleid: %s, version %s...' %
        (name, bundleid, versionstring))

    # find installed apps that match this item by name or bundleid
    appdata = info.filtered_app_data()
    appinfo = [item for item in appdata
               if (item['path'] and
                   (item['bundleid'] == bundleid or
                    (name and item['name'] == name)))]

    if not appinfo:
        # No matching apps found
        display.display_debug1(
            '\tFound no matching applications on the startup disk.')
        return ITEM_NOT_PRESENT

    # sort highest version first
    try:
        appinfo.sort(key=itemgetter('version'), reverse=True)
    except KeyError:
        # some item did not have a version key
        pass

    # iterate through matching applications
    end_result = ITEM_NOT_PRESENT
    for item in appinfo:
        if 'name' in item:
            display.display_debug2('\tFound name: \t %s', item['name'])
        display.display_debug2('\tFound path: \t %s', item['path'])
        display.display_debug2(
            '\tFound CFBundleIdentifier: \t %s', item['bundleid'])
        # create a test_app item with our found path
        test_app = {}
        test_app.update(app)
        test_app['path'] = item['path']
        compare_result = compare_bundle_version(test_app)
        if compare_result in (VERSION_IS_THE_SAME, VERSION_IS_HIGHER):
            return compare_result
        elif compare_result == VERSION_IS_LOWER:
            end_result = VERSION_IS_LOWER

    # didn't find an app with the same or higher version
    if end_result == VERSION_IS_LOWER:
        display.display_debug1(
            'An older version of this application is present.')
    return end_result


def compare_bundle_version(item):
    """Compares a bundle version passed item dict.

    Returns  0 if the bundle isn't installed
                or doesn't have valid Info.plist
            -1 if it's older
             1 if the version is the same
             2 if the version is newer

    Raises utils.Error if there's an error in the input
    """
    # look for an Info.plist inside the bundle
    filepath = os.path.join(item['path'], 'Contents', 'Info.plist')
    if not os.path.exists(filepath):
        display.display_debug1('\tNo Info.plist found at %s', filepath)
        filepath = os.path.join(item['path'], 'Resources', 'Info.plist')
        if not os.path.exists(filepath):
            display.display_debug1('\tNo Info.plist found at %s', filepath)
            return ITEM_NOT_PRESENT

    display.display_debug1('\tFound Info.plist at %s', filepath)
    # just let comparePlistVersion do the comparison
    saved_path = item['path']
    item['path'] = filepath
    compare_result = compare_plist_version(item)
    item['path'] = saved_path
    return compare_result


def compare_plist_version(item):
    """Gets the version string from the plist at path and compares versions.

    Returns  0 if the plist isn't installed
            -1 if it's older
             1 if the version is the same
             2 if the version is newer

    Raises utils.Error if there's an error in the input
    """
    version_comparison_key = item.get(
        'version_comparison_key', 'CFBundleShortVersionString')
    if 'path' in item and version_comparison_key in item:
        versionstring = item[version_comparison_key]
        filepath = item['path']
        minupvers = item.get('minimum_update_version')
    else:
        raise utils.Error('Missing plist path or version!')

    display.display_debug1('\tChecking %s for %s %s...',
                           filepath, version_comparison_key, versionstring)
    if not os.path.exists(filepath):
        display.display_debug1('\tNo plist found at %s', filepath)
        return ITEM_NOT_PRESENT

    try:
        plist = FoundationPlist.readPlist(filepath)
    except FoundationPlist.NSPropertyListSerializationException:
        display.display_debug1('\t%s may not be a plist!', filepath)
        return ITEM_NOT_PRESENT
    if not hasattr(plist, 'get'):
        display.display_debug1(
            'plist not parsed as NSCFDictionary: %s', filepath)
        return ITEM_NOT_PRESENT

    if 'version_comparison_key' in item:
        # specific key has been supplied,
        # so use this to determine installed version
        display.display_debug1(
            '\tUsing version_comparison_key %s', version_comparison_key)
        installedvers = pkgutils.getVersionString(
            plist, version_comparison_key)
    else:
        # default behavior
        installedvers = pkgutils.getVersionString(plist)
    if installedvers:
        display.display_debug1(
            '\tInstalled item has version %s', installedvers)
        if minupvers:
            if compare_versions(installedvers, minupvers) < 1:
                display.display_debug1(
                    '\tVersion %s too old < %s', installedvers, minupvers)
                return ITEM_NOT_PRESENT
        compare_result = compare_versions(installedvers, versionstring)
        results = ['older', 'not installed?!', 'the same', 'newer']
        display.display_debug1(
            '\tInstalled item is %s.', results[compare_result + 1])
        return compare_result
    else:
        display.display_debug1('\tNo version info in %s.', filepath)
        return ITEM_NOT_PRESENT


def filesystem_item_exists(item):
    """Checks to see if a filesystem item exists.

    If item has md5checksum attribute, compares on disk file's checksum.

    Returns 0 if the filesystem item does not exist on disk,
    Returns 1 if the filesystem item exists and the checksum matches
                (or there is no checksum)
    Returns -1 if the filesystem item exists but the checksum does not match.

    Broken symlinks are OK; we're testing for the existence of the symlink,
    not the item it points to.

    Raises utils.Error is there's a problem with the input.
    """
    if 'path' in item:
        filepath = item['path']
        display.display_debug1('Checking existence of %s...', filepath)
        if os.path.lexists(filepath):
            display.display_debug2('\tExists.')
            if 'md5checksum' in item:
                storedchecksum = item['md5checksum']
                ondiskchecksum = munkihash.getmd5hash(filepath)
                display.display_debug2('Comparing checksums...')
                if storedchecksum == ondiskchecksum:
                    display.display_debug2('Checksums match.')
                    return ITEM_MATCHES
                # storedchecksum != ondiskchecksum
                display.display_debug2(
                    'Checksums differ: expected %s, got %s',
                    storedchecksum, ondiskchecksum)
                return ITEM_DOES_NOT_MATCH
            # 'md5checksum' not in item
            return ITEM_MATCHES
        # not os.path.lexists(filepath)
        display.display_debug2('\tDoes not exist.')
        return ITEM_NOT_PRESENT
    # not 'path' in item
    raise utils.Error('No path specified for filesystem item.')


def compare_item_version(item):
    '''Compares an installs_item with what's on the startup disk.
    Wraps other comparison functions.

    For applications, bundles, and plists:
    Returns 0 if the item isn't installed
                or doesn't have valid Info.plist
            -1 if it's older
             1 if the version is the same
             2 if the version is newer

    For other filesystem items:
    Returns 0 if the filesystem item does not exist on disk,
            1 if the filesystem item exists and the checksum matches
                (or there is no checksum)
           -1 if the filesystem item exists but the checksum does not match.
    '''
    if not 'VersionString' in item and 'CFBundleShortVersionString' in item:
        # Ensure that 'VersionString', if not present, is populated
        # with the value of 'CFBundleShortVersionString' if present
        item['VersionString'] = item['CFBundleShortVersionString']
    itemtype = item.get('type')
    if itemtype == 'application':
        return compare_application_version(item)
    if itemtype == 'bundle':
        return compare_bundle_version(item)
    if itemtype == 'plist':
        return compare_plist_version(item)
    if itemtype == 'file':
        return filesystem_item_exists(item)
    raise utils.Error('Unknown installs item type: %s' % itemtype)


def compare_receipt_version(item):
    """Determines if the given package is already installed.

    Args:
      item: dict with packageid; a 'com.apple.pkg.ServerAdminTools' style id

    Returns  0 if the receipt isn't present
            -1 if it's older
             1 if the version is the same
             2 if the version is newer

    Raises utils.Error if there's an error in the input
    """
    if item.get('optional'):
        # receipt has been marked as optional, so it doesn't matter
        # if it's installed or not. Return 1
        # only check receipts not marked as optional
        display.display_debug1(
            'Skipping %s because it is marked as optional',
            item.get('packageid', item.get('name')))
        return VERSION_IS_THE_SAME
    installedpkgs = pkgutils.getInstalledPackages()
    if 'packageid' in item and 'version' in item:
        pkgid = item['packageid']
        vers = item['version']
    else:
        raise utils.Error('Missing packageid or version info!')

    display.display_debug1('Looking for package %s, version %s', pkgid, vers)
    installedvers = installedpkgs.get(pkgid)
    if installedvers:
        return compare_versions(installedvers, vers)
    # not installedvers
    display.display_debug1('\tThis package is not currently installed.')
    return ITEM_NOT_PRESENT


def get_installed_version(item_plist):
    """Attempts to determine the currently installed version of an item.

    Args:
      item_plist: pkginfo plist of an item to get the version for.

    Returns:
      String version of the item, or 'UNKNOWN' if unable to determine.

    """
    for receipt in item_plist.get('receipts', []):
        # look for a receipt whose version matches the pkginfo version
        if compare_versions(receipt.get('version', 0),
                            item_plist['version']) == 1:
            pkgid = receipt['packageid']
            display.display_debug2(
                'Using receipt %s to determine installed version of %s',
                pkgid, item_plist['name'])
            return pkgutils.getInstalledPackageVersion(pkgid)

    # try using items in the installs array to determine version
    install_items_with_versions = [item
                                   for item in item_plist.get('installs', [])
                                   if 'CFBundleShortVersionString' in item]
    for install_item in install_items_with_versions:
        # look for an installs item whose version matches the pkginfo version
        if compare_versions(install_item['CFBundleShortVersionString'],
                            item_plist['version']) == 1:
            if install_item['type'] == 'application':
                name = install_item.get('CFBundleName')
                bundleid = install_item.get('CFBundleIdentifier')
                display.display_debug2(
                    'Looking for application %s, bundleid %s',
                    name, install_item.get('CFBundleIdentifier'))
                try:
                    # check default location for app
                    filepath = os.path.join(install_item['path'],
                                            'Contents', 'Info.plist')
                    plist = FoundationPlist.readPlist(filepath)
                    return plist.get('CFBundleShortVersionString', 'UNKNOWN')
                except (KeyError,
                        FoundationPlist.NSPropertyListSerializationException):
                    # that didn't work, fall through to the slow way
                    appinfo = []
                    appdata = info.app_data()
                    if appdata:
                        for ad_item in appdata:
                            if bundleid and ad_item['bundleid'] == bundleid:
                                appinfo.append(ad_item)
                            elif name and ad_item['name'] == name:
                                appinfo.append(ad_item)

                    maxversion = '0.0.0.0.0'
                    for ai_item in appinfo:
                        if ('version' in ai_item and
                                compare_versions(
                                    ai_item['version'], maxversion) == 2):
                            # version is higher
                            maxversion = ai_item['version']
                    return maxversion
            elif install_item['type'] == 'bundle':
                display.display_debug2(
                    'Using bundle %s to determine installed version of %s',
                    install_item['path'], item_plist['name'])
                filepath = os.path.join(install_item['path'],
                                        'Contents', 'Info.plist')
                try:
                    plist = FoundationPlist.readPlist(filepath)
                    return plist.get('CFBundleShortVersionString', 'UNKNOWN')
                except FoundationPlist.NSPropertyListSerializationException:
                    pass
            elif install_item['type'] == 'plist':
                display.display_debug2(
                    'Using plist %s to determine installed version of %s',
                    install_item['path'], item_plist['name'])
                try:
                    plist = FoundationPlist.readPlist(install_item['path'])
                    return plist.get('CFBundleShortVersionString', 'UNKNOWN')
                except FoundationPlist.NSPropertyListSerializationException:
                    pass
    # if we fall through to here we have no idea what version we have
    return 'UNKNOWN'


if __name__ == '__main__':
    print('This is a library of support tools for the Munki Suite.')

File: code/cli/munki/shared/updatecheck/compare.swift
//
//  compare.swift
//  munki
//
//  Created by Greg Neagle on 8/19/24.
//  Copyright 2025-2026 The Munki Project. All rights reserved.
//

import Foundation

private let display = DisplayAndLog.main

enum MunkiComparisonResult: Int {
    case older = -1
    case notPresent = 0
    case same = 1
    case newer = 2

    static let different = MunkiComparisonResult.older
}

let comparisonResultDescriptions = [
    "older",
    "not present",
    "the same",
    "newer",
]

/// Compares two version numbers to one another.
/// Returns MunkiComparisonResult (one of .older, .same, .newer)
func compareVersions(_ thisVersion: String, _ thatVersion: String) -> MunkiComparisonResult {
    if MunkiVersion(thisVersion) < MunkiVersion(thatVersion) {
        return .older
    }
    if MunkiVersion(thisVersion) == MunkiVersion(thatVersion) {
        return .same
    }
    return .newer
}

func compareUsingVersionScript(_ item: PlistDict) async -> MunkiComparisonResult {
    let itemVersion = item["version"] as? String ?? "0"
    let results = await runEmbeddedScriptAndReturnResults(name: "version_script", pkginfo: item, suppressError: true)
    if results.exitcode != 0 {
        // treat an error as .notPresent
        display.debug1("\tVersion script error \(results.exitcode): \(results.error)")
        return .notPresent
    }
    let installedVersion = results.output.trimmingCharacters(in: .whitespacesAndNewlines)
    if installedVersion.isEmpty {
        display.debug1("\tNo version info returned from version script.")
        return .notPresent
    }
    display.debug1("\tVersion script returned: \(installedVersion)")
    let comparisonResult = compareVersions(installedVersion, itemVersion)
    display.debug1("\tInstalled item is \(comparisonResultDescriptions[comparisonResult.rawValue + 1])")
    return comparisonResult
}

/// Gets the version string from the plist at path and compares versions with plistItem
/// May throw a MunkiError if there's an error in the input
func comparePlistVersion(_ item: PlistDict) throws -> MunkiComparisonResult {
    let versionComparisonKey = item["version_comparison_key"] as? String ?? "CFBundleShortVersionString"
    guard let path = item["path"] as? String,
          let versionString = item[versionComparisonKey] as? String
    else {
        throw MunkiError("Missing plist path or version!")
    }
    let minimumUpdateVersion = item["minimum_update_version"] as? String
    display.debug1("\tChecking \(path) for \(versionComparisonKey) \(versionString)...")
    if !pathExists(path) {
        display.debug1("\tNo file found at \(path)")
        return .notPresent
    }
    guard let plist = try? readPlist(fromFile: path) as? PlistDict else {
        display.debug1("\t\(path) can't be read as a plist!")
        return .notPresent
    }
    let installedVersion: String
    if item.keys.contains("version_comparison_key") {
        // specific key has been supplied,
        // so use this to determine installed version
        display.debug1("\tUsing version_comparison_key \(versionComparisonKey)")
        installedVersion = getVersionString(plist: plist, key: versionComparisonKey)
    } else {
        // just use default behavior
        installedVersion = getVersionString(plist: plist)
    }
    if installedVersion.isEmpty {
        display.debug1("\tNo version info in \(path).")
        return .notPresent
    }
    display.debug1("\tInstalled item has version \(installedVersion)")
    if let minimumUpdateVersion {
        if compareVersions(installedVersion, minimumUpdateVersion) == .older {
            display.debug1("\tInstalled version \(installedVersion) is too old to update (must be \(minimumUpdateVersion) or later)")
            return .notPresent
        }
    }
    let comparisonResult = compareVersions(installedVersion, versionString)
    display.debug1("\tInstalled item is \(comparisonResultDescriptions[comparisonResult.rawValue + 1])")
    return comparisonResult
}

/// Compares bundle described in bundleItem with what is actually on-disk.
/// May throw a MunkiError if there's an error in the input
func compareBundleVersion(_ bundleItem: PlistDict) throws -> MunkiComparisonResult {
    if let path = bundleItem["path"] as? NSString {
        var infoPlistPath = path.appendingPathComponent("Contents/Info.plist")
        if !pathExists(infoPlistPath) {
            display.debug1("\tNo Info.plist found at \(infoPlistPath)")
            infoPlistPath = path.appendingPathComponent("Resources/Info.plist")
            if !pathExists(infoPlistPath) {
                display.debug1("\tNo Info.plist found at \(infoPlistPath)")
                return .notPresent
            }
        }
        display.debug1("\tFound Info.plist at \(infoPlistPath)")
        var plistItem = bundleItem
        plistItem["path"] = infoPlistPath
        return try comparePlistVersion(plistItem)
    }
    return .notPresent
}

/// Compares app described in appItem with what is actually on-disk.
/// Checks the given path if it's available,
/// otherwise uses LaunchServices and/or Spotlight to look for the app
/// May throw a MunkiError if there's an error in the input
func compareApplicationVersion(_ appItem: PlistDict) throws -> MunkiComparisonResult {
    if let path = appItem["path"] as? String {
        if !pathExists(path) {
            display.debug2("Application is not present at \(path).")
            return .notPresent
        }
        let infoPlistPath = (path as NSString).appendingPathComponent("Contents/Info.plist")
        if !pathExists(infoPlistPath) {
            display.debug2("Application Info.plist does not exist.")
            return .notPresent
        }
        return try compareBundleVersion(appItem)
    }
    // no 'path' in appItem
    display.debug2("No path provided for application item.")
    let bundleName = appItem["CFBundleName"] as? String ?? ""
    let bundleID = appItem["CFBundleIdentifier"] as? String ?? ""
    let versionComparisonKey = appItem["version_comparison_key"] as? String ?? "CFBundleShortVersionString"
    let versionString = appItem[versionComparisonKey] as? String ?? ""

    if bundleName.isEmpty, bundleID.isEmpty {
        // we have no path, no bundleName, no bundle identifier. Error!
        throw MunkiError("No path, bundle name or bundle identifier was specified!")
    }

    display.debug1("Looking for application \(bundleName) with bundleid: \(bundleID), version \(versionString)...")

    // find installed apps that match this item by name or bundleid
    let appData = filteredAppData()
    var appInfo = appData.filter {
        $0["path"] != nil &&
            (($0["bundleid"] ?? "<nobundleid>") == bundleID ||
                ($0["name"] ?? "<nobundlename>") == bundleName)
    }

    if appInfo.isEmpty {
        // no matching apps found
        display.debug1("\tFound no matching applications on the startup disk.")
        return .notPresent
    }

    // sort highest version first
    appInfo.sort {
        MunkiVersion($0["version"] ?? "") > MunkiVersion($1["version"] ?? "")
    }

    // iterate through matching apps
    var endResult = MunkiComparisonResult.notPresent
    for item in appInfo {
        display.debug2("\tFound name: \(item["name"] ?? "<none>")")
        display.debug2("\tFound path: \(item["path"] ?? "<none>")")
        display.debug2("\tFound CFBundleIdentifier: \(item["bundleid"] ?? "<none>")")
        // create a test item to use for comparison
        var testItem = appItem
        testItem["path"] = item["path"]
        let compareResult = try compareBundleVersion(testItem)
        if compareResult == .same || compareResult == .newer {
            return compareResult
        }
        if compareResult == .older {
            endResult = .older
        }
    }

    // didn't find an app with the same or higher version
    if endResult == .older {
        display.debug1("An older version of this application is present.")
    }
    return endResult
}

/// Returns the status of the local filesystem item as compared to
/// the passed-in dictionary
///
/// If item has md5checksum attribute, compares on disk file's checksum.
///
/// Throws a MunkiError if there's a problem with the input
func filesystemItemStatus(_ item: PlistDict) throws -> MunkiComparisonResult {
    guard let filepath = item["path"] as? String else {
        throw MunkiError("No path specified for filesystem item")
    }
    display.debug1("Checking existence of \(filepath)...")
    if pathExists(filepath) {
        display.debug2("\tExists.")
        if let expectedChecksum = item["md5checksum"] as? String {
            display.debug2("Comparing checksums...")
            let onDiskChecksum = md5hash(file: filepath)
            if onDiskChecksum == expectedChecksum {
                display.debug2("Checksums match.")
                return .same
            }
            // onDiskChecksum != expectedChecksum
            display.debug2("Checksums differ: expected \(expectedChecksum), found \(onDiskChecksum)")
            return .different
        }
        // md5checksum not in item, but item is present
        return .same
    }
    // path doesn't exist
    return .notPresent
}

/// Compares an installs_item with what's on the startup disk.
/// Wraps other comparison functions.
/// Returns a MunkiComparisonResult
/// Can throw MunkiError on bad input
func compareItem(_ item: PlistDict) throws -> MunkiComparisonResult {
    guard let type = item["type"] as? String else {
        throw MunkiError("Item type was not defined")
    }
    switch type {
    case "application":
        return try compareApplicationVersion(item)
    case "bundle":
        return try compareBundleVersion(item)
    case "plist":
        return try comparePlistVersion(item)
    case "file":
        return try filesystemItemStatus(item)
    default:
        throw MunkiError("Unknown or unsupported installs item type: \(type)")
    }
}

/// Determines if the given package is already installed.
/// Input: dict with receipt info
/// Returns a MunkiComparisonResult
/// Can throw MunkiError on bad input
func compareReceipt(_ item: PlistDict) async throws -> MunkiComparisonResult {
    if item["optional"] as? Bool ?? false {
        // receipt has been marked as optional, so it doesn't matter
        // if it's installed or not. Return .same
        // only check receipts not marked as optional
        return .same
    }
    guard let pkgid = item["packageid"] as? String,
          let receiptVersion = item["version"] as? String
    else {
        throw MunkiError("Receipt item is missing packageid or version info!")
    }
    display.debug1("Looking for package \(pkgid), version \(receiptVersion)")
    let installedPkgs = await getInstalledPackages()
    if let installedVersion = installedPkgs[pkgid] {
        return compareVersions(installedVersion, receiptVersion)
    }
    // no installedVersion
    display.debug1("\tThis package is not currently installed.")
    return .notPresent
}

/// Checks a package id against the receipts to determine if a package is already installed.
/// Returns the version string of the installed pkg if it exists, or an empty string if it does not
func getInstalledPackageVersion(_ pkgid: String) -> String? {
    let results = runCLI(
        "/usr/sbin/pkgutil", arguments: ["--pkg-info-plist", pkgid]
    )
    if results.exitcode == 0 {
        guard let receipt = (try? readPlist(fromString: results.output)) as? PlistDict else {
            display.debug2("Unable to parse output from pkgutil")
            return nil
        }
        guard let foundpkgid = receipt["pkgid"] as? String else {
            display.debug2("No pkgid in pkgutil output")
            return nil
        }
        guard let foundversion = receipt["pkg-version"] as? String else {
            display.debug2("No version in pkgutil output")
            return nil
        }
        if foundpkgid == pkgid {
            display.debug1(
                "\tThis machine has \(pkgid), version \(foundversion)")
            return foundversion
        }
    }
    // This package does not appear to be currently installed
    display.debug1("\tThis machine does not have \(pkgid)")
    return nil
}

/// Attempts to determine the currently installed version of an item.
///
/// Args:
/// pkginfo: pkginfo plist of an item to get the version for.
///
/// Returns:
/// String version of the item, or 'UNKNOWN' if unable to determine
func getInstalledVersion(_ pkginfo: PlistDict) -> String {
    func versionFromPlist(_ path: String) -> String? {
        do {
            if let plist = try readPlist(fromFile: path) as? PlistDict {
                return plist["CFBundleShortVersionString"] as? String ?? "UNKNOWN"
            }
            display.debug2("plist \(path) in wrong format")
        } catch {
            display.debug2("plist \(path) read error: \(error.localizedDescription)")
        }
        return nil
    }

    let itemName = pkginfo["name"] as? String ?? "<unknown>"
    let itemVersion = pkginfo["version"] as? String ?? "-1"

    // try receipts
    let receipts = pkginfo["receipts"] as? [PlistDict] ?? []
    for receipt in receipts {
        // look for a receipt whose version matches the pkginfo version
        // (there is no guarantee at all that a receipt/pkg version matches
        //  the version of the software installed by the package)
        if let pkgid = receipt["packageid"] as? String,
           let receiptVersion = receipt["version"] as? String,
           compareVersions(receiptVersion, itemVersion) == .same
        {
            display.debug2("Using receipt \(pkgid) to determine installed version of \(itemName)")
            if let pkgVersion = getInstalledPackageVersion(pkgid) {
                return pkgVersion
            }
        }
    }
    // try using installs_array
    if let installItems = pkginfo["installs"] as? [PlistDict] {
        // filter out items that don't actually have version info
        let installItemsWithVersions = installItems.filter {
            $0["CFBundleShortVersionString"] != nil
        }
        for installItem in installItemsWithVersions {
            // look for an installs item whose version matches the pkginfo version
            guard let itemType = installItem["type"] as? String,
                  let installItemVersion = installItem["CFBundleShortVersionString"] as? String
            else {
                continue
            }
            if compareVersions(installItemVersion, itemVersion) != .same {
                continue
            }
            let path = installItem["path"] as? String ?? "<unknown>"
            if itemType == "application" {
                display.debug2("Using application \(path) to determine installed version of \(itemName)")
                if path != "<unknown>" {
                    let infopath = (path as NSString).appendingPathComponent("Contents/Info.plist")
                    if let installedVersion = versionFromPlist(infopath) {
                        return installedVersion
                    }
                }
                // find installed apps that match this item by name or bundleid
                let bundleName = installItem["CFBundleName"] as? String ?? ""
                let bundleID = installItem["CFBundleIdentifier"] as? String ?? ""
                let appData = filteredAppData()
                var appInfo = appData.filter {
                    $0["path"] != nil &&
                        (($0["bundleid"] ?? "<nobundleid>") == bundleID ||
                            ($0["name"] ?? "<nobundlename>") == bundleName)
                }

                if appInfo.isEmpty {
                    // no matching apps found
                    display.debug1("\tFound no matching applications on the startup disk.")
                    continue
                }

                // sort highest version first
                appInfo.sort {
                    MunkiVersion($0["version"] ?? "") > MunkiVersion($1["version"] ?? "")
                }
                if let installedVersion = appInfo[0]["version"] {
                    return installedVersion
                }
            } else if itemType == "bundle" {
                display.debug2("Using bundle \(path) to determine installed version of \(itemName)")
                let infopath = (path as NSString).appendingPathComponent("Contents/Info.plist")
                if let installedVersion = versionFromPlist(infopath) {
                    return installedVersion
                }
            } else if itemType == "plist" {
                display.debug2("Using plist \(path) to determine installed version of \(itemName)")
                if let installedVersion = versionFromPlist(path) {
                    return installedVersion
                }
            }
        }
    }
    // if we fall through to here we have no idea what version we have
    return "UNKNOWN"
}
